src/Template/TemplateVariableContainer.php

```diff
@@ -0,0 +1,126 @@
+ * after the class itself) in order to aggregate template variables prior to
+ * rendering. The class that renders a template (generally a handler, but
+ * potentially middleware) can then pull this container and use it to seed the
+ * template variables.
+ *
+ * Middleware that needs to populate one or more template variables can do the
+ * following:
+ *
+ * <code>
+ * $container = $request->getAttribute(
+ *     TemplateVariableContainer::class,
+ *     new TemplateVariableContainer()
+ * );
+ *
+ * // Populate a single variable:
+ * $container->set('user', $user);
+ *
+ * // Populate several variables:
+ * $container->merge([
+ *     'user'  => $user,
+ *     'roles' => $user->getRoles(),
+ * ]);
+ *
+ * return $handler->handle($request->withAttribute(
+ *     TemplateVariableContainer::class,
+ *     $container
+ * ));
+ * </code>
+ *
+ * In a handler or middleware that renders a template, pull the container, and
+ * use its mergeForTemplate() method when calling render():
+ *
+ * <code>
+ * $container = $request->getAttribute(
+ *     TemplateVariableContainer::class,
+ *     new TemplateVariableContainer()
+ * );
+ *
+ * $content = $this->renderer->render(
+ *     'some::template',
+ *     $container->mergeForTemplate([
+ *         'local' => 'value',
+ *     ])
+ * );
+ * </code>
+ */
+class TemplateVariableContainer implements Countable
+{
+    /**
+     * @var array<string, mixed>
+     */
+    private $variables = [];
+
+    public function count() : int
+    {
+        return count($this->variables);
+    }
+
+    /**
+     * @return null|mixed Returns null if $key does not exist in container;
+     *     otherwise, returns value associated with $key
+     */
+    public function get(string $key)
+    {
+        return $this->variables[$key] ?? null;
+    }
+
+    public function has(string $key) : bool
+    {
+        return array_key_exists($key, $this->variables);
+    }
+
+    public function set(string $key, $value) : void
+    {
+        $this->variables[$key] = $value;
+    }
+
+    public function unset(string $key) : void
+    {
+        unset($this->variables[$key]);
+    }
+
+    /**
+     * Merge an array of values into the container.
+     *
+     * Use this method to populate many values in the container at once.
+     *
+     * This method will overwrite any existing value with the same key with the
+     * new value if it occurs in $values.
+     */
+    public function merge(array $values) : void
+    {
+        $this->variables = array_merge($this->variables, $values);
+    }
+
+    /**
+     * Merge a set of values with those in the container, and return the result.
+     *
+     * Use this method to merge handler-specific template values with those in
+     * the container in order to pass the result to the renderer's `render()`
+     * method.
+     */
+    public function mergeForTemplate(array $values) : array
+    {
+        return array_merge($this->variables, $values);
```

Comment:
Shouldn't be a recursive merge?

Reply:
I wasn't sure if there was a good use case for that. Generally speaking, you assign discrete values to templates: scalars, possibly objects, occasionally arrays. I'm not sure it would make sense to do a recursive merge for arrays, as you would generally be replacing the value at the top-level assignment.

Reply:
@weierophinney You're right ;)

BTW: I would want require your branch through composer with commit ref but it seem I cannot access it. I miss something?

Reply:
It's because the patch is submitted via my fork, and not as a branch on this repo (I submit PRs the same as anybody else does! :smile:). As such, you will need to add a `repositories` section to your `composer.json`:

```json
"repositories": [
  {"type": "vcs", "url": "https://github.com/weierophinney/zend-expressive-helpers.git"}
]
```

Then, where you `require` zend-expressive-helpers in your `composer.json`, update it as follows:

```json
"zendframework/zend-expressive-json": "dev-feature/template-variable-container as 5.2.0",
```

Then run `composer update`.

Reply:
@weierophinney 

For the repository thing, I was aware of ;) The problem: I was not able to find your fork through your github profile ;) I must be blind...

Reply:
@weierophinney 

Lol, you have a typo I think in your comment ;)

```
"zendframework/zend-expressive-json": "dev-feature/template-variable-container as 5.2.0",
```
JSON ;)

Reply:
> you have a typo I think in your comment

fixed.